File: packages/shared/src/presets.ts
import type { ExportOptions, ObjectTypeOption } from "./dto";

export type ComponentTemplate = {
  name: string;
  required: boolean;
  materialSlots: string[];
};

export type MaterialTemplate = {
  name: string;
  roughnessDefault: number;
  normalStrengthDefault: number;
};

export type PresetDefinition = {
  id: Exclude<ObjectTypeOption, "auto-detect" | `studio:${string}`>;
  label: string;
  components: ComponentTemplate[];
  materials: MaterialTemplate[];
  defaultExportOptions: ExportOptions;
};

const defaultExportOptions: ExportOptions = {
  units: "cm",
  pivot: "floor-center",
  smoothing: true,
  fixBackfaces: true,
  keepMaterialsSeparated: true,
  keepComponentsSeparated: true,
  includePbrMaps: true,
  textureSize: "2k",
  includeLightweight: true,
};

export const BUILTIN_PRESETS: PresetDefinition[] = [
  {
    id: "chair-stool",
    label: "Chair/Stool",
    components: [
      { name: "OBJ_Seat", required: true, materialSlots: ["MAT_Wood_Primary", "MAT_Fabric_Seat"] },
      { name: "OBJ_Backrest", required: true, materialSlots: ["MAT_Fabric_Seat", "MAT_Wood_Primary"] },
      { name: "OBJ_Legs", required: true, materialSlots: ["MAT_Metal_Detail", "MAT_Wood_Primary"] },
      { name: "OBJ_Frame", required: true, materialSlots: ["MAT_Wood_Primary", "MAT_Metal_Detail"] },
      { name: "OBJ_Hardware", required: false, materialSlots: ["MAT_Metal_Detail"] },
    ],
    materials: [
      { name: "MAT_Wood_Primary", roughnessDefault: 0.62, normalStrengthDefault: 0.25 },
      { name: "MAT_Fabric_Seat", roughnessDefault: 0.84, normalStrengthDefault: 0.35 },
      { name: "MAT_Metal_Detail", roughnessDefault: 0.28, normalStrengthDefault: 0.18 },
    ],
    defaultExportOptions,
  },
  {
    id: "table-desk",
    label: "Table/Desk",
    components: [
      { name: "OBJ_Top", required: true, materialSlots: ["MAT_Wood_Top", "MAT_Stone_Top"] },
      { name: "OBJ_Legs", required: true, materialSlots: ["MAT_Wood_Base", "MAT_Metal_Detail"] },
      { name: "OBJ_Frame", required: true, materialSlots: ["MAT_Wood_Base", "MAT_Metal_Detail"] },
      { name: "OBJ_Drawers", required: false, materialSlots: ["MAT_Wood_Base"] },
      { name: "OBJ_Hardware", required: false, materialSlots: ["MAT_Metal_Detail"] },
    ],
    materials: [
      { name: "MAT_Wood_Top", roughnessDefault: 0.56, normalStrengthDefault: 0.22 },
      { name: "MAT_Wood_Base", roughnessDefault: 0.6, normalStrengthDefault: 0.2 },
      { name: "MAT_Metal_Detail", roughnessDefault: 0.3, normalStrengthDefault: 0.2 },
      { name: "MAT_Stone_Top", roughnessDefault: 0.48, normalStrengthDefault: 0.16 },
    ],
    defaultExportOptions,
  },
  {
    id: "sofa-textile",
    label: "Sofa/Textile",
    components: [
      { name: "OBJ_Frame", required: true, materialSlots: ["MAT_Wood_Base", "MAT_Metal_Detail"] },
      { name: "OBJ_Cushions_Seat", required: true, materialSlots: ["MAT_Fabric_Primary"] },
      { name: "OBJ_Cushions_Back", required: true, materialSlots: ["MAT_Fabric_Secondary", "MAT_Fabric_Primary"] },
      { name: "OBJ_Arms", required: true, materialSlots: ["MAT_Fabric_Primary"] },
      { name: "OBJ_Legs", required: true, materialSlots: ["MAT_Wood_Base", "MAT_Metal_Detail"] },
      { name: "OBJ_Hardware", required: false, materialSlots: ["MAT_Metal_Detail"] },
    ],
    materials: [
      { name: "MAT_Fabric_Primary", roughnessDefault: 0.9, normalStrengthDefault: 0.38 },
      { name: "MAT_Fabric_Secondary", roughnessDefault: 0.84, normalStrengthDefault: 0.33 },
      { name: "MAT_Wood_Base", roughnessDefault: 0.62, normalStrengthDefault: 0.22 },
      { name: "MAT_Metal_Detail", roughnessDefault: 0.32, normalStrengthDefault: 0.18 },
    ],
    defaultExportOptions,
  },
  {
    id: "lighting",
    label: "Lighting",
    components: [
      { name: "OBJ_Shade", required: true, materialSlots: ["MAT_Fabric_Shade", "MAT_Glass_Primary"] },
      { name: "OBJ_Body", required: true, materialSlots: ["MAT_Metal_Body"] },
      { name: "OBJ_Cable", required: false, materialSlots: ["MAT_Metal_Body"] },
      { name: "OBJ_Bulb", required: true, materialSlots: ["MAT_Emissive"] },
      { name: "OBJ_Hardware", required: false, materialSlots: ["MAT_Metal_Body"] },
    ],
    materials: [
      { name: "MAT_Glass_Primary", roughnessDefault: 0.12, normalStrengthDefault: 0.1 },
      { name: "MAT_Metal_Body", roughnessDefault: 0.26, normalStrengthDefault: 0.16 },
      { name: "MAT_Fabric_Shade", roughnessDefault: 0.78, normalStrengthDefault: 0.3 },
      { name: "MAT_Emissive", roughnessDefault: 0.1, normalStrengthDefault: 0.05 },
    ],
    defaultExportOptions,
  },
  {
    id: "decor",
    label: "Decor",
    components: [
      { name: "OBJ_Main", required: true, materialSlots: ["MAT_Main_Surface"] },
      { name: "OBJ_Base", required: false, materialSlots: ["MAT_Base"] },
      { name: "OBJ_Details", required: false, materialSlots: ["MAT_Detail"] },
    ],
    materials: [
      { name: "MAT_Main_Surface", roughnessDefault: 0.55, normalStrengthDefault: 0.2 },
      { name: "MAT_Base", roughnessDefault: 0.62, normalStrengthDefault: 0.16 },
      { name: "MAT_Detail", roughnessDefault: 0.4, normalStrengthDefault: 0.2 },
    ],
    defaultExportOptions,
  },
  {
    id: "rug-curtain",
    label: "Rug/Curtain",
    components: [
      { name: "OBJ_Main", required: true, materialSlots: ["MAT_Fabric_Primary"] },
      { name: "OBJ_Folds", required: false, materialSlots: ["MAT_Fabric_Secondary", "MAT_Fabric_Primary"] },
    ],
    materials: [
      { name: "MAT_Fabric_Primary", roughnessDefault: 0.88, normalStrengthDefault: 0.4 },
      { name: "MAT_Fabric_Secondary", roughnessDefault: 0.8, normalStrengthDefault: 0.34 },
    ],
    defaultExportOptions,
  },
  {
    id: "divider",
    label: "Divider",
    components: [
      { name: "OBJ_Main", required: true, materialSlots: ["MAT_Main_Surface"] },
      { name: "OBJ_Base", required: false, materialSlots: ["MAT_Base"] },
      { name: "OBJ_Details", required: false, materialSlots: ["MAT_Detail"] },
    ],
    materials: [
      { name: "MAT_Main_Surface", roughnessDefault: 0.58, normalStrengthDefault: 0.22 },
      { name: "MAT_Base", roughnessDefault: 0.64, normalStrengthDefault: 0.18 },
      { name: "MAT_Detail", roughnessDefault: 0.42, normalStrengthDefault: 0.2 },
    ],
    defaultExportOptions,
  },
];

const presetById = new Map(BUILTIN_PRESETS.map((preset) => [preset.id, preset]));

export function getPresetByType(type: Exclude<ObjectTypeOption, "auto-detect" | `studio:${string}`>) {
  return presetById.get(type) ?? presetById.get("decor")!;
}

export function normalizeStrictComponentOrder(
  preset: PresetDefinition,
  presentComponentNames: string[]
): ComponentTemplate[] {
  const lookup = new Set(presentComponentNames);
  return preset.components.filter((component) => component.required || lookup.has(component.name));
}

export function defaultMaterialTemplatesForPreset(
  preset: PresetDefinition,
  allowedNames?: string[]
): MaterialTemplate[] {
  if (!allowedNames || allowedNames.length === 0) return preset.materials;
  const names = new Set(allowedNames);
  return preset.materials.filter((material) => names.has(material.name));
}

export function resolveAutoDetectPreset(input: {
  objectType: ObjectTypeOption;
  reconstructionMode: "auto" | "rigid" | "organic";
  fileHints?: string[];
  materialHints?: string[];
}): Exclude<ObjectTypeOption, "auto-detect" | `studio:${string}`> {
  if (input.objectType !== "auto-detect" && !input.objectType.startsWith("studio:")) {
    return input.objectType as Exclude<ObjectTypeOption, "auto-detect" | `studio:${string}`>;
  }

  const mergedHints = [...(input.fileHints ?? []), ...(input.materialHints ?? [])]
    .join(" ")
    .toLowerCase();

  if (/fabric|textile|sofa|cushion|curtain|rug/.test(mergedHints)) return "sofa-textile";
  if (/lamp|shade|light|bulb/.test(mergedHints)) return "lighting";
  if (/table|desk/.test(mergedHints)) return "table-desk";
  if (/chair|stool/.test(mergedHints)) return "chair-stool";
  if (/divider|panel/.test(mergedHints)) return "divider";
  if (input.reconstructionMode === "organic") return "sofa-textile";
  if (input.reconstructionMode === "rigid") return "chair-stool";
  return "decor";
}

File: apps/service/app/presets.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class MaterialTemplate:
    name: str
    roughness: float
    normal_strength: float


@dataclass(frozen=True)
class ComponentTemplate:
    name: str
    required: bool
    material_slots: tuple[str, ...]


@dataclass(frozen=True)
class PresetDefinition:
    id: str
    label: str
    components: tuple[ComponentTemplate, ...]
    materials: tuple[MaterialTemplate, ...]


PRESET_MAP: dict[str, PresetDefinition] = {
    "chair-stool": PresetDefinition(
        id="chair-stool",
        label="Chair/Stool",
        components=(
            ComponentTemplate("OBJ_Seat", True, ("MAT_Wood_Primary", "MAT_Fabric_Seat")),
            ComponentTemplate("OBJ_Backrest", True, ("MAT_Fabric_Seat", "MAT_Wood_Primary")),
            ComponentTemplate("OBJ_Legs", True, ("MAT_Metal_Detail", "MAT_Wood_Primary")),
            ComponentTemplate("OBJ_Frame", True, ("MAT_Wood_Primary", "MAT_Metal_Detail")),
            ComponentTemplate("OBJ_Hardware", False, ("MAT_Metal_Detail",)),
        ),
        materials=(
            MaterialTemplate("MAT_Wood_Primary", 0.62, 0.25),
            MaterialTemplate("MAT_Fabric_Seat", 0.84, 0.35),
            MaterialTemplate("MAT_Metal_Detail", 0.28, 0.18),
        ),
    ),
    "table-desk": PresetDefinition(
        id="table-desk",
        label="Table/Desk",
        components=(
            ComponentTemplate("OBJ_Top", True, ("MAT_Wood_Top", "MAT_Stone_Top")),
            ComponentTemplate("OBJ_Legs", True, ("MAT_Wood_Base", "MAT_Metal_Detail")),
            ComponentTemplate("OBJ_Frame", True, ("MAT_Wood_Base", "MAT_Metal_Detail")),
            ComponentTemplate("OBJ_Drawers", False, ("MAT_Wood_Base",)),
            ComponentTemplate("OBJ_Hardware", False, ("MAT_Metal_Detail",)),
        ),
        materials=(
            MaterialTemplate("MAT_Wood_Top", 0.56, 0.22),
            MaterialTemplate("MAT_Wood_Base", 0.6, 0.2),
            MaterialTemplate("MAT_Metal_Detail", 0.3, 0.2),
            MaterialTemplate("MAT_Stone_Top", 0.48, 0.16),
        ),
    ),
    "sofa-textile": PresetDefinition(
        id="sofa-textile",
        label="Sofa/Textile",
        components=(
            ComponentTemplate("OBJ_Frame", True, ("MAT_Wood_Base", "MAT_Metal_Detail")),
            ComponentTemplate("OBJ_Cushions_Seat", True, ("MAT_Fabric_Primary",)),
            ComponentTemplate("OBJ_Cushions_Back", True, ("MAT_Fabric_Secondary", "MAT_Fabric_Primary")),
            ComponentTemplate("OBJ_Arms", True, ("MAT_Fabric_Primary",)),
            ComponentTemplate("OBJ_Legs", True, ("MAT_Wood_Base", "MAT_Metal_Detail")),
            ComponentTemplate("OBJ_Hardware", False, ("MAT_Metal_Detail",)),
        ),
        materials=(
            MaterialTemplate("MAT_Fabric_Primary", 0.9, 0.38),
            MaterialTemplate("MAT_Fabric_Secondary", 0.84, 0.33),
            MaterialTemplate("MAT_Wood_Base", 0.62, 0.22),
            MaterialTemplate("MAT_Metal_Detail", 0.32, 0.18),
        ),
    ),
    "lighting": PresetDefinition(
        id="lighting",
        label="Lighting",
        components=(
            ComponentTemplate("OBJ_Shade", True, ("MAT_Fabric_Shade", "MAT_Glass_Primary")),
            ComponentTemplate("OBJ_Body", True, ("MAT_Metal_Body",)),
            ComponentTemplate("OBJ_Cable", False, ("MAT_Metal_Body",)),
            ComponentTemplate("OBJ_Bulb", True, ("MAT_Emissive",)),
            ComponentTemplate("OBJ_Hardware", False, ("MAT_Metal_Body",)),
        ),
        materials=(
            MaterialTemplate("MAT_Glass_Primary", 0.12, 0.1),
            MaterialTemplate("MAT_Metal_Body", 0.26, 0.16),
            MaterialTemplate("MAT_Fabric_Shade", 0.78, 0.3),
            MaterialTemplate("MAT_Emissive", 0.1, 0.05),
        ),
    ),
    "decor": PresetDefinition(
        id="decor",
        label="Decor",
        components=(
            ComponentTemplate("OBJ_Main", True, ("MAT_Main_Surface",)),
            ComponentTemplate("OBJ_Base", False, ("MAT_Base",)),
            ComponentTemplate("OBJ_Details", False, ("MAT_Detail",)),
        ),
        materials=(
            MaterialTemplate("MAT_Main_Surface", 0.55, 0.2),
            MaterialTemplate("MAT_Base", 0.62, 0.16),
            MaterialTemplate("MAT_Detail", 0.4, 0.2),
        ),
    ),
    "rug-curtain": PresetDefinition(
        id="rug-curtain",
        label="Rug/Curtain",
        components=(
            ComponentTemplate("OBJ_Main", True, ("MAT_Fabric_Primary",)),
            ComponentTemplate("OBJ_Folds", False, ("MAT_Fabric_Secondary", "MAT_Fabric_Primary")),
        ),
        materials=(
            MaterialTemplate("MAT_Fabric_Primary", 0.88, 0.4),
            MaterialTemplate("MAT_Fabric_Secondary", 0.8, 0.34),
        ),
    ),
    "divider": PresetDefinition(
        id="divider",
        label="Divider",
        components=(
            ComponentTemplate("OBJ_Main", True, ("MAT_Main_Surface",)),
            ComponentTemplate("OBJ_Base", False, ("MAT_Base",)),
            ComponentTemplate("OBJ_Details", False, ("MAT_Detail",)),
        ),
        materials=(
            MaterialTemplate("MAT_Main_Surface", 0.58, 0.22),
            MaterialTemplate("MAT_Base", 0.64, 0.18),
            MaterialTemplate("MAT_Detail", 0.42, 0.2),
        ),
    ),
}


def resolve_auto_preset(object_type: str, reconstruction_mode: str, hints: list[str]) -> str:
    if object_type in PRESET_MAP:
        return object_type

    merged = " ".join(hints).lower()
    if any(term in merged for term in ["fabric", "textile", "sofa", "cushion", "rug", "curtain"]):
        return "sofa-textile"
    if any(term in merged for term in ["lamp", "light", "shade", "bulb"]):
        return "lighting"
    if any(term in merged for term in ["table", "desk"]):
        return "table-desk"
    if any(term in merged for term in ["chair", "stool"]):
        return "chair-stool"
    if reconstruction_mode == "organic":
        return "sofa-textile"
    if reconstruction_mode == "rigid":
        return "chair-stool"
    return "decor"
